### PR TITLE
Improve `check_thread_leak` output

### PR DESCRIPTION
**Pull request by gjoseph92**
This should aid in debugging #6796.

This now gives output like:
```
2 thread(s) were leaked from test

------ Call stack of leaked thread 1/2: <Thread(Thread-3, started 123145522995200)> ------
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 912, in _bootstrap
        self._bootstrap_inner()
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 954, in _bootstrap_inner
        self.run()
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 892, in run
        self._target(*self._args, **self._kwargs)
  File "/Users/gabe/dev/distributed/distributed/tests/test_utils_test.py", line 770, in <lambda>
        t2 = threading.Thread(target=lambda: (event.wait(), "two"))
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 574, in wait
        signaled = self._cond.wait(timeout)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 312, in wait
        waiter.acquire()

------ Call stack of leaked thread 2/2: <Thread(Thread-4, started 123145539784704)> ------
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 912, in _bootstrap
        self._bootstrap_inner()
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 954, in _bootstrap_inner
        self.run()
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 892, in run
        self._target(*self._args, **self._kwargs)
  File "/Users/gabe/dev/distributed/distributed/tests/test_utils_test.py", line 772, in <lambda>
        t3 = threading.Thread(target=lambda: (event.wait(), "three"))
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 574, in wait
        signaled = self._cond.wait(timeout)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/threading.py", line 312, in wait
        waiter.acquire()
```

Instead of:
```
>               assert False, (bad_thread, call_stacks)
E               AssertionError: (<Thread(asyncio_4, started daemon 123145657024512)>, ['  File "/Users/runner/miniconda3/envs/dask-distributed/lib/pyt...ibuted/lib/python3.10/concurrent/futures/thread.py", line 81, in _worker
E                 \twork_item = work_queue.get(block=True)
E                 '])
E               assert False
```
which was truncating the traceback and not giving much useful information.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/tests/test_utils_test.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 import pytest
 import yaml
+from _pytest.outcomes import Failed
```
Comment (jacobtomlinson) [inline distributed/tests/test_utils_test.py:19]: Using private classes seems a little dangerous.

**Review by jacobtomlinson — COMMENTED** [review]
This looks neat! Would be useful to see when this is happening.

File: distributed/tests/test_utils_test.py (hunk 1 above)
Comment (graingert) [reply distributed/tests/test_utils_test.py:19]: It's made public in `pytest.fail.Exception`

File: distributed/utils_test.py [hunk 2]
```diff
@@ -1750,9 +1750,17 @@ def check_thread_leak():
...
-            bad_thread = bad_threads[0]
-            call_stacks = profile.call_stack(sys._current_frames()[bad_thread.ident])
-            assert False, (bad_thread, call_stacks)
+            lines: list[str] = [f"{len(bad_threads)} thread(s) were leaked from test\n"]
+            for i, thread in enumerate(bad_threads, 1):
+                lines.append(
+                    f"------ Call stack of leaked thread {i}/{len(bad_threads)}: {thread} ------"
+                )
+                lines.append(
+                    "".join(profile.call_stack(sys._current_frames()[thread.ident]))
```
Comment (graingert) [inline distributed/utils_test.py:1759]: I think it's better to call sys._current_frames once and then delete it when finished

File: distributed/tests/test_utils_test.py (hunk 1 above)
Comment (gjoseph92) [reply distributed/tests/test_utils_test.py:19]: I didn't like that either. Cool!

**Comment by gjoseph92** [comment]
- https://github.com/dask/distributed/pull/6817
- https://github.com/dask/distributed/issues/6796 (what this is meant to help diagnose)

**Pull request merged**